src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/TANInputViewController/TanInputView.swift

```diff
@@ -0,0 +1,236 @@
+//
+// 🦠 Corona-Warn-App
+//
+
+import UIKit
+
+class TanInputView: UIControl, UIKeyInput {
+
+	// MARK: - Init
+
+	init(
+		frame: CGRect,
+		viewModel: TanInputViewModel
+	) {
+		self.viewModel = viewModel
+		super.init(frame: frame)
+
+		setup()
+		setupAccessibility()
+	}
+
+	required init?(coder: NSCoder) {
+		fatalError("init(coder:) has not been implemented")
+	}
+
+	// MARK: - Overrides
+
+	override var canBecomeFirstResponder: Bool {
+		true
+	}
+
+	// MARK: - Protocol UIKeyInput
+
+	var hasText: Bool {
+		return !viewModel.text.isEmpty
+	}
+
+	func insertText(_ text: String) {
+		guard text != "\n" else {
+			viewModel.submitTan()
+			return
+		}
+
+		for character in text.trimmingCharacters(in: .whitespacesAndNewlines).map({ $0.uppercased() }) {
+			guard !viewModel.isValid && !viewModel.isInputBlocked else {
+				return
+			}
+
+			let label = inputLabels[viewModel.text.count]
+			// updates a single boxed label
+			label.text = "\(character)"
+			label.isValid = character.rangeOfCharacter(from: characterSet) != nil
+			viewModel.isInputBlocked = !label.isValid
+			viewModel.appendCharacter(character)
+		}
+	}
+
+	func deleteBackward() {
+		guard !viewModel.text.isEmpty else {
+				return
+			}
+		viewModel.isInputBlocked = false
+		viewModel.deletLastCharacter()
+		inputLabels[viewModel.text.count].clear()
+	}
+
+	// MARK: - Public
+
+	// MARK: - Internal
```

Comment:
We don't have these, so why adding them here?

Reply:
removed